presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java

```diff
@@ -0,0 +1,482 @@
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        return SimplePlanRewriter.rewriteWith(new Optimizer(idAllocator, symbolAllocator, metadata),
+                plan,
+                Optional.empty());
+    }
+
+    private static class Optimizer
+            extends SimplePlanRewriter<Optional<AggregateInfo>>
+    {
+        private final PlanNodeIdAllocator idAllocator;
+        private final SymbolAllocator symbolAllocator;
+        private final Metadata metadata;
+
+        private Optimizer(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Metadata metadata)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
+            this.metadata = requireNonNull(metadata, "metadata is null");
+        }
+
+        @Override
+        public PlanNode visitAggregation(AggregationNode node, RewriteContext<Optional<AggregateInfo>> context)
+        {
+            // optimize if and only if
+            // some aggregation functions have a distinct mask symbol
+            // and if not all aggregation functions on same distinct mask symbol (this case handled by SingleDistinctOptimizer)
+            Set<Symbol> masks = ImmutableSet.copyOf(node.getMasks().values());
+            if (masks.size() != 1 || node.getMasks().size() == node.getAggregations().size()) {
+                return context.defaultRewrite(node, Optional.empty());
+            }
+
+            AggregateInfo aggregateInfo = new AggregateInfo(node.getGroupBy(),
+                    Iterables.getOnlyElement(masks),
+                    node.getAggregations(),
+                    node.getFunctions());
+
+            PlanNode source = context.rewrite(node.getSource(), Optional.of(aggregateInfo));
+
+            // Change aggregate node to do second aggregation, handles this part of optimized plan mentioned above:
+            //          SELECT a1, a2,..., an, arbitrary(f1 if group = 0 else null),...., arbitrary(fm if group = 0 else null), F(c if group = 1 else null)
+            ImmutableMap.Builder<Symbol, FunctionCall> aggregations = ImmutableMap.builder();
+            ImmutableMap.Builder<Symbol, Signature> functions = ImmutableMap.builder();
+            for (Map.Entry<Symbol, FunctionCall> entry : node.getAggregations().entrySet()) {
+                FunctionCall functionCall = entry.getValue();
+                if (entry.getValue().isDistinct()) {
+                    aggregations.put(entry.getKey(), new FunctionCall(functionCall.getName(),
+                            functionCall.getWindow(),
+                            false,
+                            ImmutableList.of(Iterables.getOnlyElement(aggregateInfo.getNewDistinctAggregateSymbols().values()).toSymbolReference())));
```

Comment:
Will change it, this was left as map  because I originally thought of writing this for multiple distinct aggregations.